src/bwt_util.rs

```diff
@@ -1,3 +1,111 @@
+use bitvec::prelude::*;
+use itertools::Itertools;
+use std::clone::Clone;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::hash::Hash;
+
+/// implementation of the mergeIter function as described in Holt & McMillan 2014
+/// Takes the current state of the interleave bitvector, the two burrows wheeler
+/// transform slices, and a mapping from character to number of lexigraphically
+/// smaller characters in the BWTs provided and does a round of interleave
+/// resolution
+pub fn pairwise_merge_iter<T: Ord + Hash + Clone + Debug>(
+    current_interleave: &BitVec<u64, Msb0>,
+    bwt0: &[T],
+    bwt1: &[T],
+    offsets: &HashMap<T, usize>,
+) -> BitVec<u64, Msb0> {
+    // initial conditions
+    let mut next_interleave = bitvec![u64, Msb0; 0; current_interleave.len()];
+    let mut current_pos0 = 0usize;
+    let mut current_pos1 = 0usize;
+
+    let mut temp_index = offsets.clone();
+
+    // iterate through each bit
+    for b in current_interleave.iter() {
+        let c = if *b {
+            let v = bwt0[current_pos0].clone();
+            current_pos0 += 1;
+            v
+        } else {
+            let v = bwt1[current_pos1].clone();
+            current_pos1 += 1;
+            v
+        };
+        // copy b into the next position for c
+        let next_position = temp_index
+            .get_mut(&c)
+            .expect("Unknown character encountered in merging BWTs");
+        next_interleave.set(*next_position, *b);
+        // update the tempIndex to match the FM-index
+        *next_position += 1;
+    }
+
+    next_interleave
+}
+
+/// Determines the offset hashmap for a provided slice of BWTs.
+/// The offset hashmap is the mapping from a character to the number of
+/// lexigraphically lesser characters present in the combined string provided.
+fn generate_offset_hashmap<T: Ord + Hash + Clone + Debug>(bwts: &[&[T]]) -> HashMap<T, usize> {
+    let mut num_occurrences: HashMap<T, usize> = HashMap::new();
+    for &bwt in bwts.iter() {
+        for c in bwt {
+            if let Some(i) = num_occurrences.get_mut(c) {
+                *i += 1;
+            } else {
+                num_occurrences.insert(c.clone(), 1);
+            }
+        }
+    }
+    let ordered_chars = num_occurrences.keys().sorted().cloned().collect::<Vec<T>>();
+    let mut total = 0usize;
+    let mut offset_map = HashMap::with_capacity(num_occurrences.len());
+    for c in ordered_chars {
+        offset_map.insert(c.clone(), total);
+        total += num_occurrences[&c];
+    }
+    offset_map
+}
+
+/// Merges two Burrows-Wheeler transformed slices into a new, merged Burrows-Wheeler
+/// slice.
+pub fn pairwise_bwt_merge<T: Ord + Hash + Clone + Debug>(bwt0: &[T], bwt1: &[T]) -> Vec<T> {
```

Comment:
This one in particular will need a more in-depth comment block with `# Arguments` and `# Example` tags (see `naive_bwt(...)` for example typical docs).  It would be good to have at least the Arguments tag for the other two added functions as well.